### DISCUSSION TITLE
better support for java and jdtls

### DISCUSSION THREAD

**Discussion by SoPat712** [LunarVim/LunarVim#3200] [Q&A]
### Problem description

When trying to use JDTLS, I get this error
![image](https://user-images.githubusercontent.com/30350506/186014404-caf90830-bce8-4832-a87c-a6d91a7bff03.png)

I checked :LspLog and saw this error

![image](https://user-images.githubusercontent.com/30350506/186014444-1c1b9420-8c5f-4762-8ff8-9f26f8eca6ae.png)

### LunarVim version

master-47e4e5b

### Neovim version (>= 0.7)

NVIM v0.7.2

### Operating system/version

macOS 11.6.6

### Affected language servers

jdtls

### Steps to reproduce

1. Create java file
2. Run :LspInstall jdtls
3. Run :LspStart jdtls
4. See the error in the top right

### Actual behavior

![image](https://user-images.githubusercontent.com/30350506/186014274-6a9d76e7-8393-4151-8ea9-28c07bff9b97.png)
and 
![image](https://user-images.githubusercontent.com/30350506/186014334-8b2e9921-00a7-4920-ba37-72571e57989d.png)

### Expected behavior

Proper language processing and code completion, etc.

### support info

LspInfo:

```console
  Language client log: /Users/joshp/.cache/lvim/lsp.log
 Detected filetype:   java
 
 0 client(s) attached to this buffer: 
 
 Other clients that match the filetype: java
 
 Config: jdtls
 	Refer to :h lspconfig-root-detection for help.
 	filetypes:         java
 	root directory:    Not found. Searched for: { -- Single-module projects { 'build.xml', -- Ant 'pom.xml', -- Maven 'settings.gradle', -- Gradle 'settings.gradle.kts', -- Gradle }, -- Multi-module projects { 'build.gradle', 'build.gradle.kts' }, } or vim.fn.getcwd().
 	cmd:               java -Declipse.application=org.eclipse.jdt.ls.core.id1 -Dosgi.bundles.defaultStartLevel=4 -Declipse.product=org.eclipse.jdt.ls.core.product --add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/sun.nio.fs=ALL-UNNAMED -jar /Users/joshp/.local/share/nvim/lsp_servers/jdtls/plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar -configuration /Users/joshp/.local/share/nvim/lsp_servers/jdtls/config_mac -data /Users/joshp/workspace/ProgrammingAssigment2 -XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Djava.import.generatesMetadataFilesAtProjectRoot=false -Xmx1G -Xms100m
 	cmd is executable: true (checked by nvim-lsp-installer)
 	autostart:         true
 	custom handlers:   textDocument/rename, textDocument/codeAction, workspace/applyEdit, language/status
 
 Configured servers list: jdtls
```

LvimInfo:

```console
                                   
                                       __                          _    ___         
                                      / /   __  ______  ____ _____| |  / (_)___ ___ 
                                     / /   / / / / __ \/ __ `/ ___/ | / / / __ `__ \
                                    / /___/ /_/ / / / / /_/ / /   | |/ / / / / / / /
                                   /_____/\__,_/_/ /_/\__,_/_/    |___/_/_/ /_/ /_/ 
                                   
                                   
                                   Buffer info
                                   * filetype:                java
                                   * bufnr:                   11
                                   * treesitter status:       active
                                   
                                   Active client(s)
                                   
                                   
                                   
                                   Formatters info
                                   * Active: 
                                   * Supported: [astyle, clang_format, google_java_format, uncrustify]
                                   
                                   Linters info
                                   * Active: 
                                   * Supported: semgrep
                                   
                                   Code actions info
                                   * Active: 
```

### logs

```console
[ERROR][2022-08-22 16:37:23] .../vim/lsp/rpc.lua:420	"rpc"	"java"	"stderr"	"WARNING: Using incubator modules: jdk.incubator.jpackage, jdk.incubator.foreign\n"
```

**Comment by erlonbie** [comment]
I have the same issue, but not specifically using LunarVim, just normal neovim.

**Comment by kylo252** [comment]
This seems like an issue with lspconfig or the server itself.

We could use a streamlined configuration for nvim-jdtls if anyone has a working example, maybe something like this https://github.com/ChristianChiarulli/nvim/blob/b8fc718ab9f592df9488ba92948e46605d1cc5be/ftplugin/java.lua

**Comment by SoPat712** [comment]
I used this [guide](https://youtu.be/0q_MKUynUck) and got jdtls running.

**Comment by lvimuser** [comment]
> We could use a streamlined configuration for nvim-jdtls if anyone has a working example, maybe something like this https://github.com/ChristianChiarulli/nvim/blob/b8fc718ab9f592df9488ba92948e46605d1cc5be/ftplugin/java.lua

I have a working configuration which doesn't use the binary, instead fetching the config from nvim lsp installer. Thing is, that configuration is missing from mason-lspconfig, so I'm not sure what to do with it (we can just pull it from nvim lsp installer, but idk). 

We could also easily add debug support once we add two packages to mason.

------------------

`user.jdtls`:

```lua
local M = {}

local user_opts = {
  name = "jdtls",
  use_lombok_agent = true,
  on_attach = function(...)
    require("jdtls.dap").setup_dap_main_class_configs()
    require("jdtls").setup_dap({ hotcodereplace = "auto" })
    require("jdtls.setup").add_commands()

    require("lvim.lsp").common_on_attach(...)
  end,
}

local _, server = require("nvim-lsp-installer.servers").get_server("jdtls")
local root_dir = require("nvim-lsp-installer.server").get_server_root_path("jdtls")
local installer_opts = server:get_default_options()

M.config = {}
M.setup = function()
  if not vim.tbl_isempty(M.config) then
    installer_opts.on_new_config(M.config, root_dir)
    return M.config
  end

  local jdtls = require "jdtls"
  local extendedCapabilities = jdtls.extendedClientCapabilities
  extendedCapabilities.resolveAdditionalTextEditsSupport = true

  local lvim_common_opts = require("lvim.lsp").get_common_opts()
  local opts = vim.tbl_deep_extend("force", lvim_common_opts, installer_opts, user_opts, {
    settings = {
      extendedClientCapabilities = extendedCapabilities,
    },
  })

  -- https://github.com/williamboman/nvim-lsp-installer/issues/501#issuecomment-1050664641
  -- fill in the correct 'cmd'
  installer_opts.on_new_config(opts, root_dir)

  M.config = opts
  return M.config
end

return M
```

`ftplugin/java.lua`:

```lua
local status_ok, jdtls = pcall(require, "jdtls")
if not status_ok then
  return
end

jdtls.start_or_attach(require("user.jdtls").setup())
```

**Comment by lvimuser** [comment]
For the error, see https://github.com/mfussenegger/nvim-jdtls/discussions/247
I fixed with `rm -rf ~/workspace/<project_name>`

**Comment by kylo252** [comment]
> Thing is, that configuration is missing from mason-lspconfig, so I'm not sure what to do with it (we can just pull it from nvim lsp installer, but idk).

@lvimuser, maybe they should be re-added then if it's necessary, or maybe to lspconfig directly.

Otherwise, I would just implement it manually in `lvim/lsp/providers/jdtls.lua`, we can get the install path easily from mason-registry although I'm not really sure it's relevant since it's already on `$PATH`

---

here's `java.lua` by the author of nvim-jdtls, I think @ChristianChiarulli had actually built upon it: https://github.com/mfussenegger/dotfiles/blob/c9d1efcf738bc924fce8a138d68ce84621b4ea07/vim/.config/nvim/ftplugin/java.lua

**Comment by JoTerrance** [comment]
I had same problem with vanilla neovim(not tested in lunarvim yes), i found the problem was jdk version, in my case it requires java 17, hope it helps

**Comment by lvimuser** [comment]
Debug support is easier now after https://github.com/williamboman/mason.nvim/pull/482 (see the first comment for an example).

**Comment by kylo252** [comment]
check https://github.com/LunarVim/starter.lvim/tree/java-ide for examples
